share/wake/lib/system/sources.wake

```diff
@@ -58,6 +58,58 @@ export def source (file: String): Result Path Error =
         x, Nil = Pass x
         _      = unreachable "a quoted RegExp can never match more than one distinct String"
 
+# Allows claiming of a file so long as the file is within
+# wakeroot. Keep in mind that if another job produces this
+# file, claim will fail. Likewise if you claim a file then
+# any job that produces this file will fail. Claim should
+# be thought of as a "virtual" job and wake demands that
+# all paths be produced by exactly one job, virtual or
+# otherwise.
+#
+# For source files please use `source`. For files outside
+# of wakeroot please use `claimFileAsPath` or `claimFileAsPathIn`.
+# `claim` is good for claiming files of previous builds that
+# are not being used again this build.
+#
+# Parameters:
+#  - `file`: The string giving the path to the file to be claimed
+#
+# Example:
+# ```
+#   claim "build/my_flow/my_step/previous-output.txt" # Claims the file
+# ```
+export def claim (file: String): Result Path Error =
+    # Define the actual virtual job that produces the Path
+    def raw_claim file =
+        def get_modtime file = prim "get_modtime"
+        def time = get_modtime file
+        if time == -1 then
+            failWithError "{file}: files does not exist. Maybe you haven't previously built this file?"
+        else
+            # Its important that the command here is distinct from what `source` would use
+            # so that claim and source do not overlap.
+            makeExecPlan ("<claim>", str time, file, Nil) Nil
```

Comment:
Yes. You cannot both claim and source a file. You can however claim a file twice (which I had not intended initially but after speaking with Wes I guess thats how `source` works).

Reply:
I feel that the error message is probably going to confuse a user then, they wont view source and claim as jobs

Reply:
I'll add that to that query, that query should not work that way most likely. What its trying to do I guess is identify virtual jobs. That indicates to me that we need a marker in the database for virtual jobs results.

Reply:
I think I can use `prim "sources"` to make this work, it doesn't hash anything so I can add a check for this here; I'll throw it up real quick. One sec.

Reply:
Ok I added some handling for that, here's the latest manual testing (I should add these as unit tests huh?)

```
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "..test.txt"'
Pass (Path "..test.txt")
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "../..test.txt"'
Fail (Error "claim cannot be used on files outside of wake's root. Consider using claimFileAsPath instead." Nil)
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "/scratch/jakee/test.txt"'
Fail (Error "claim cannot be used on files outside of wake's root. Consider using claimFileAsPath instead." Nil)
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "/scratch/jakee/wake_build/test.txt"'
Fail (Error "/scratch/jakee/wake_build/test.txt: Is a source file, please use `source` or `sources` instead of `claim`" Nil)
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "test.txt"'
Fail (Error "test.txt: Is a source file, please use `source` or `sources` instead of `claim`" Nil)
jakee@gamma06:/scratch/jakee/wake_build$ touch test2.txt
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "test2.txt"'
Pass (Path "test2.txt")
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "/scratch/jakee/wake_build/test2.txt"'
Pass (Path "test2.txt")
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "../test.txt"'
Fail (Error "claim cannot be used on files outside of wake's root. Consider using claimFileAsPath instead." Nil)
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "foo/../../test.txt"'
Fail (Error "claim cannot be used on files outside of wake's root. Consider using claimFileAsPath instead." Nil)
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "foo/../test.txt"'
Fail (Error "foo/../test.txt: Is a source file, please use `source` or `sources` instead of `claim`" Nil)
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "foo/../test2.txt"'
Pass (Path "test2.txt")
jakee@gamma06:/scratch/jakee/wake_build$ ./bin/wake -x 'claim "test.txt"'
Fail (Error "test.txt: Is a source file, please use `source` or `sources` instead of `claim`" Nil)
```

With this `source` and `claim` will never succeed on each others turf so to speak. We can have a discussion later about a possible unification. I know that having to `git add` everything is annoying for people. I haven't heard the merits of using it quite.

Reply:
seems good to me